File: 7회차/198_house_robber/198_house_robber.c
#include <stdio.h>
#include <stdlib.h>

// House Robber 함수
int rob(int* nums, int numsSize) {
    if (numsSize == 0) return 0;           // 빈 입력 처리
    if (numsSize == 1) return nums[0];     // 집이 하나인 경우

    int prev2 = nums[0];                   // dp[i-2]: 첫 번째 집만 털 경우
    int prev1 = nums[0] > nums[1] ? nums[0] : nums[1]; // dp[i-1]: 첫 두 집 중 최댓값

    for (int i = 2; i < numsSize; i++) {
        int curr = prev1 > (nums[i] + prev2) ? prev1 : (nums[i] + prev2); // dp[i] 계산
        prev2 = prev1;               // dp[i-2] 갱신
        prev1 = curr;                // dp[i-1] 갱신
    }

    return prev1;                    // 최종 결과 반환
}

// 테스트 함수
int main() {
    int nums1[] = {1, 2, 3, 1};
    int numsSize1 = sizeof(nums1) / sizeof(nums1[0]);
    printf("Maximum amount robbed: %d\n", rob(nums1, numsSize1)); // Expected: 4

    int nums2[] = {2, 7, 9, 3, 1};
    int numsSize2 = sizeof(nums2) / sizeof(nums2[0]);
    printf("Maximum amount robbed: %d\n", rob(nums2, numsSize2)); // Expected: 12

    int nums3[] = {2};
    int numsSize3 = sizeof(nums3) / sizeof(nums3[0]);
    printf("Maximum amount robbed: %d\n", rob(nums3, numsSize3)); // Expected: 2

    return 0;
}

File: 7회차/198_house_robber/198_house_robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:  # 빈 입력 처리
            return 0
        if len(nums) == 1:  # 집이 하나일 경우
            return nums[0]

        dp = nums[:]
        dp[1] = max(nums[0], nums[1])  # 두 번째 집까지 고려

        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 1], nums[i] + dp[i - 2])  # 이전 최대와 현재 집을 포함한 최대 비교

        return dp[-1]  # 마지막 값 반환